### DISCUSSION TITLE
Resubscribe on Error Message

### DISCUSSION THREAD

**Discussion by zhixinwen** [enisdenjo/graphql-ws#435] [Q&A] [answered]
The server return an error type message to client, and I want to resubscribe upon error (websocket is not closed by server).

How do I subscribe again? The major challenge is that in theory, there can be multiple subscriptions going on in the websocket, and the only identification I can get is `id`.

Do I have to store `id` -> `subscription query` in a map when I start a subscription, and look up the `id` in the map for query when error occurs? Is there a better way for it?

**Comment by enisdenjo** [comment, marked as answer]
Maybe I am not understanding your question, but why not literally re-subscribe on error? Roughly:
```js
import { createClient } from 'graphql-ws';

const client = createClient({
  url: 'ws://unstable:4000/graphql',
});

(async () => {
  for (;;) {
    try {
      await new Promise((resolve, reject) => {
        client.subscribe(
          {
            query: 'subscription { iMightFail }',
          },
          {
            next: (val) => {
              console.log(val); // or emit the value however you want
            },
            error: reject,
            complete: resolve,
          },
        );
      });
      // subscription has completed successfully, break the loop to avoid subscribing again
      break;
    } catch (err) {
      // subscription received an error message, simply continuing the loop will resubscribe
      continue;
    }
  }
})();
```

**Reply by zhixinwen** [reply]
@aaronbuxbaum does this work for our case?

**Reply by AaronBuxbaum** [reply]
IMO, no -- this design requires every customer to know how to use it, and requires every use-case to implement the functionality independently. You could build a utility function and share it between customers, but again, it requires every customer to be aware of very unusual management, and for clients to manage it

It does _work_ in the sense that yes, the functionality works, but it is a suboptimal solution vs the standard: retry in the server and then throw error or code if enough retries fail. At that point, the client can respond however it makes sense to that client

To be clear, this is the best you can do on any client with a server that doesn't retry subscriptions, it's not a problem with graphql-ws itself
